staging/src/k8s.io/kubectl/pkg/cmd/cmd.go

```diff
@@ -522,6 +520,33 @@ func NewKubectlCommand(in io.Reader, out, err io.Writer) *cobra.Command {
 	matchVersionKubeConfigFlags := cmdutil.NewMatchVersionFlags(kubeConfigFlags)
 	matchVersionKubeConfigFlags.AddFlags(cmds.PersistentFlags())
 
+	// Hook before and after Run initialize and write profiles to disk,
+	// respectively. Also, PreRun adds kubectl command headers to be
+	// subsequently added to each request. See SIG CLI KEP 859:
+	//   https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/859-kubectl-headers
+	crt := &genericclioptions.CommandHeaderRoundTripper{}
+	cmds.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		rest.SetDefaultWarningHandler(warningHandler)
+		// Stores kubectl commands X-Headers here for subsequent addition
+		// to each Request.
+		crt.ParseCommandHeaders(cmd, args)
+		// Generated header with new request id here and put in map
+		return initProfiling()
+	}
```

Comment:
I'd move this entire block below inside the env lookup code. Doing it here means you're unnecessarily calling `ParseCommandHeaders` on every call. I'd suggest leaving `PersistentPreRunE` setup as it was before and inside the below condition do something like:
```golang
if _, exists := os.LookupEnv(kubectlCmdHeaders); exists {
	crt := &genericclioptions.CommandHeaderRoundTripper{}
	existingPreRunE := cmds.PersistentPreRunE
	cmds.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
		crt.ParseCommandHeaders(cmd, args)
		return existingPreRunE(cmd, args)
	}

	kubeConfigFlags.WrapConfigFn = func(c *rest.Config) *rest.Config {
		c.Wrap(func(rt http.RoundTripper) http.RoundTripper {
			crt.Delegate = rt
			return crt
		})
		return c
	}
}
```

Reply:
I'd probably advise moving this entire setup block to a separate function, for readbility :smile: